Implement backend level fallback for c10

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29682 backend fallback test
* #29681 better boxing
* #29337 Hide the OperatorKernel* argument from the stack based kernel API
* #29330 OperatorHandle::callBoxed/callUnboxed
* #29214 Test cases for backend fallback kernels
* #29201 Add a OperatorHandle argument to boxed kernels
* **#28494 Implement backend level fallback for c10**
* #28314 Simplify c10 dispatcher

Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

smessmer committed:
[wip] Implement backend level fallback for c10

Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

bwasti commented:
include test?  at least to demonstrate usage

bwasti reviewed (approved):
stamping to unblock

smessmer committed:
Update on "[wip] Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

bwasti commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -237,6 +243,8 @@ inline Return Dispatcher::callUnboxed(const OperatorHandle& op, Args... args) co
 
 template<class Return, class... Args>
 inline Return Dispatcher::callUnboxedOnly(const OperatorHandle& op, Args... args) const {
+  unused_arg_(args...);  // workaround for a false-positive warning about unused parameters in gcc 5
```
why do we need this?

smessmer replied:
gcc 5 bug, see the comment

smessmer committed:
Update on "[wip] Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit f8399e1:
* 0/15 flaky
* 1/15 broken upstream



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=f8399e1285923ab6d8e406444c8f393f113aeadd).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

smessmer committed:
Update on "[wip] Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

ezyang commented:
You should at least bring back the tests that were deleted in https://github.com/pytorch/pytorch/pull/27978

smessmer committed:
Update on "[wip] Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

dzhulgakov reviewed (commented):
Hm, why doesn't fallback function receive the schema of the operator being called? It feels quite useless without it. In the original implementation it was there: https://github.com/pytorch/pytorch/blob/062d05eab3d4530a50d9ae7e79cd0fba7c6df1f5/aten/src/ATen/test/boxed_fallback_test.cpp#L44

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -31,6 +31,7 @@ OpRegistrationListener::~OpRegistrationListener() {}
 Dispatcher::Dispatcher()
 : operators_()
 , operatorLookupTable_()
+, backendFallbackKernels_()
```
I don't think we should use the term "backend" here, since `at::Backend` is still the legacy concept and people are bound to get confused. To align with the current terminology (though I think TensorTypeId is a bad name too, right now), you should call this `perTensorTypeFallbackKernel`

mcarilli commented:
@smessmer I see people have approved this PR without the boxed_fallback_test.cpp included.  Do you still intend to restore that?  It would be very helpful to see a version of boxed_fallback_test refactored to accomodate this PR.

smessmer committed:
Update on "[wip] Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "[wip] Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

ezyang commented:
Yes please bring back the test!

smessmer commented:
@dzhulgakov @bwasti The schema argument (or actually an OperatorHandle argument) is added to this in a separate PR, see https://github.com/pytorch/pytorch/pull/29201.

@ezyang @mcarilli I'll add some test cases. Either here or in a separate PR.

smessmer committed:
Update on "Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer commented:
test cases added in https://github.com/pytorch/pytorch/pull/29214

smessmer committed:
Update on "Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

dzhulgakov commented:
@smessmer - general comment about stacked diffs. I think in this case this PR would be much easier to review if it was merged with the next two ones: backend fallback without passing the operator doesn't make much sense and unittests are best added in the same PR as the code they test. So unless the individual PRs are too huge (which they are not), I think we can split them in a fewer pieces.

smessmer committed:
Update on "Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

smessmer committed:
Update on "Implement backend level fallback for c10"


Allow a backend-level fallback kernel that is called whenever an operator doesn't have a concrete kernel for the backend.
This is needed for lazy.

Differential Revision: [D18081495](https://our.internmc.facebook.com/intern/diff/D18081495/)

[ghstack-poisoned]

MarisaKirisame commented:
suppose I want to use this to do something similar to lazy tensor. How can I get the functionschema out?
